crates/iceberg/src/arrow/reader.rs

```diff
@@ -186,4 +216,399 @@ impl ArrowReader {
+        Ok(())
+    }
+
+    fn not(&mut self, _predicate: Self::T) -> Result<Self::T> {
+        Ok(())
+    }
+
+    fn visit_always_true(&mut self) -> Result<Self::T> {
+        Ok(())
+    }
+
+    fn visit_always_false(&mut self) -> Result<Self::T> {
+        Ok(())
+    }
+
+    fn visit_unary(&mut self, predicate: &UnaryExpression<BoundReference>) -> Result<Self::T> {
+        self.bound_reference(predicate.term())?;
+        Ok(())
+    }
+
+    fn visit_binary(&mut self, predicate: &BinaryExpression<BoundReference>) -> Result<Self::T> {
+        self.bound_reference(predicate.term())?;
+        Ok(())
+    }
+
+    fn visit_set(&mut self, predicate: &SetExpression<BoundReference>) -> Result<Self::T> {
+        self.bound_reference(predicate.term())?;
+        Ok(())
+    }
+
+    fn bound_reference(&mut self, reference: &BoundReference) -> Result<Self::T> {
+        self.field_ids.push(reference.field().id);
+        Ok(())
+    }
+}
+
+/// A visitor to convert Iceberg bound predicates to Arrow predicates.
+struct PredicateConverter<'a> {
+    /// The leaf column indices used in the predicates.
+    pub columns: &'a Vec<usize>,
+    /// The projection mask for the Arrow predicates.
+    pub projection_mask: ProjectionMask,
+    /// The Parquet schema descriptor.
+    pub parquet_schema: &'a SchemaDescriptor,
+    /// The map between field id and leaf column index in Parquet schema.
+    pub column_map: &'a HashMap<i32, usize>,
+}
+
+fn get_arrow_datum(datum: &Datum) -> Result<Box<dyn ArrowDatum + Send>> {
+    match datum.literal() {
+        PrimitiveLiteral::Boolean(value) => Ok(Box::new(BooleanArray::new_scalar(*value))),
+        PrimitiveLiteral::Int(value) => Ok(Box::new(Int32Array::new_scalar(*value))),
+        PrimitiveLiteral::Long(value) => Ok(Box::new(Int64Array::new_scalar(*value))),
+        PrimitiveLiteral::Float(value) => Ok(Box::new(Float32Array::new_scalar(value.as_f32()))),
+        PrimitiveLiteral::Double(value) => Ok(Box::new(Float64Array::new_scalar(value.as_f64()))),
+        l => Err(Error::new(
+            ErrorKind::DataInvalid,
+            format!("Unsupported literal type: {:?}", l),
+        )),
+    }
+}
+
+/// Recursively get the leaf column from the record batch. Assume that the nested columns in
+/// struct is projected to a single column.
+fn get_leaf_column(column: &ArrayRef) -> std::result::Result<ArrayRef, ArrowError> {
+    match column.data_type() {
+        DataType::Struct(fields) => {
+            if fields.len() != 1 {
+                return Err(ArrowError::SchemaError(
+                    "Struct column should have only one field after projection"
+                        .parse()
+                        .unwrap(),
+                ));
+            }
+            let struct_array = column.as_any().downcast_ref::<StructArray>().unwrap();
+            get_leaf_column(struct_array.column(0))
+        }
+        _ => Ok(column.clone()),
```

Comment:
I'm not sure if `BoundReference` could be referred to nested field inside list or map in Iceberg. Assume that it is either struct or primitive now.

Reply:
I also think it doesn't support it now.